src/mapping.ts

```diff
@@ -52,75 +57,42 @@ export function handleStaked(event: Staked): void {
     account.save()
   }
 
-  const stakeAmount = event.params.amount
   stakeData.stakeType = type
   stakeData.owner = account.id
   stakeData.beneficiary = event.params.beneficiary
   stakeData.authorizer = event.params.authorizer
-  stakeData.tStake = type === "T" ? stakeAmount : BigInt.zero()
-  stakeData.nuInTStake = type === "NU" ? stakeAmount : BigInt.zero()
-  stakeData.keepInTStake = type === "KEEP" ? stakeAmount : BigInt.zero()
+  stakeData.tStake = type === "T" ? amount : BigInt.zero()
+  stakeData.nuInTStake = type === "NU" ? amount : BigInt.zero()
+  stakeData.keepInTStake = type === "KEEP" ? amount : BigInt.zero()
   stakeData.totalStaked = stakeData.tStake
     .plus(stakeData.keepInTStake)
     .plus(stakeData.nuInTStake)
   stakeData.save()
 
-  const lastEpochId = (epochCounter.count - 1).toString()
-  const lastEpoch = Epoch.load(lastEpochId)
-
-  const totalStaked = lastEpoch
-    ? lastEpoch.totalStaked.plus(event.params.amount)
-    : event.params.amount
-  let epochStakes: string[] = []
-
-  if (lastEpoch) {
-    lastEpoch.duration = blockTimestamp.minus(lastEpoch.startTime)
-    lastEpoch.save()
-
-    epochStakes = lastEpoch.stakes
-    for (let i = 0; i < epochStakes.length; i++) {
-      const epochStake = EpochStake.load(epochStakes[i])
-      const epStId = crypto
-        .keccak256(
-          ByteArray.fromUTF8(
-            epochStake!.stakeData + epochCounter.count.toString()
-          )
-        )
-        .toHexString()
-      epochStake!.id = epStId
-      epochStake!.participation = epochStake!.amount.divDecimal(
-        totalStaked.toBigDecimal()
-      )
-      epochStake!.save()
-      epochStakes[i] = epochStake!.id
-    }
-  }
-
-  const epStId = crypto
-    .keccak256(ByteArray.fromUTF8(stakeData.id + epochCounter.count.toString()))
-    .toHexString()
-  const newEpochStake = new EpochStake(epStId)
-  newEpochStake.stakeData = stakeData.id
-  newEpochStake.amount = event.params.amount
-  newEpochStake.participation = newEpochStake.amount.divDecimal(
-    totalStaked.toBigDecimal()
-  )
-  newEpochStake.save()
-
-  epochStakes.push(newEpochStake.id)
-
-  const epoch = new Epoch(epochCounter.count.toString())
-  epoch.startTime = blockTimestamp
-  epoch.totalStaked = totalStaked
+  const lastEpoch = getOrCreateLastEpoch()
+  const timestamp = event.block.timestamp
+  lastEpoch.duration = timestamp.minus(lastEpoch.timestamp)
+  lastEpoch.save()
+
+  const epochStakes = populateNewEpochStakes(lastEpoch.stakes)
```

Comment:
> TBH I don't understand why we need this and `epoch.stakes = epochstakes`. Could you please clarify a little?

Sure! Staker's ongoing rewards are calculated in function of the "participation" of the staker along the time.

Let's exemplify this:

![Staking example](https://user-images.githubusercontent.com/28781252/173553021-04fde039-e026-48b3-9f2e-3bd606aa44c5.png)

Here we have 4 epochs because we have 4 events: two staked and two unstaked.

Following the example and calculating the participation, we have:

Epoch 0: No rewards
Epoch 1: 100% of rewards for Staker 1.
Epoch 2: 66.67% of rewards for Staker 1, 33.33% of rewards for Staker 2.
Epoch 3: 100% of rewards for Staker 2.

In order to calculate the percentage value of each participant, we need to know the stakes over time.

To achieve this, when a new event is received, the list of stakes (this is, epochStakes) of the last epoch is duplicated, and depending on if the event is `staked`, `unstaked`, or `toppedup`, we will add a new staker to the list, remove it or modify the staked amount.

`populateNewEpochStakes` is the function that duplicates the list.

Reply:
I see. My only concern is iteration through the `stakes` in `Epoch` entity. I wonder if we can move that part to a script that will generate the merkle tree. I mean maybe we could find all `EpochStakes` between 2 dates by GraphQL and based on this calculate a share of the pool in this duration.